utils/parallel.go

```diff
@@ -164,3 +164,49 @@ func RunInParallel(ctx context.Context, fs []SimpleFunc) (time.Duration, error)
 	wg.Wait()
 	return time.Since(start), bigError
 }
+
+// FloatFunc is for GetInParallel.
+type FloatFunc func(ctx context.Context) (float64, error)
+
+// GetInParallel runs all functions in parallel, return is elapsed time and n error.
+func GetInParallel(ctx context.Context, fs []FloatFunc) (time.Duration, []float64, error) {
+	start := time.Now()
+	ctx, cancel := context.WithCancel(ctx)
+	var floats []float64
+
+	var wg sync.WaitGroup
+
+	var bigError error
+	var bigErrorMutex sync.Mutex
+	storeError := func(float float64, err error) {
+		bigErrorMutex.Lock()
+		defer bigErrorMutex.Unlock()
+		if bigError == nil || !errors.Is(err, context.Canceled) {
+			bigError = multierr.Combine(bigError, err)
+		}
+	}
+
+	helper := func(f FloatFunc) {
+		defer func() {
+			if thePanic := recover(); thePanic != nil {
+				storeError(0, fmt.Errorf("got panic running something in parallel: %v", thePanic))
+				cancel()
+			}
+			wg.Done()
+		}()
+		float, err := f(ctx)
+		if err != nil {
+			storeError(float, err)
+			cancel()
+		}
+		floats = append(floats, float)
```

Comment:
related to your other comment about the order of these... maybe it's obvious but I don't see why the order would change during this function. maybe something about goroutines that I'm unaware of? I assume that the loop would go through the list in order like usual

Reply:
What if one of them turns out to be much slower than the other? Maybe you've got remote parts for each motor, and the first remote part is in the middle of reconfiguration, so the second one responds more quickly. or maybe you've got some experimental, slightly-buggy modular component that becomes unresponsive when the garbage collector is running, and that causes the first motor not to respond quickly while the second one does. Your assumption that the Go runtime environment would schedule things in FIFO order sounds plausible to me, but it's not obvious that it's what actually happens: perhaps it actually schedules the most recent one first because it might still be in the L1 cache, and then does everything older than that in FIFO order. The point is just that when you run things concurrently, you have no guarantees of ordering, and need to use things like mutexes or WaitGroups to get some of that back.

Reply:
yeah that makes sense. just to be safe I added in your other suggestion with the indexes and now haven't had any issues with testing. before I added that change I did have one run that switched the two positions but should be good now